src/utils/submission.rs

```diff
@@ -31,3 +34,39 @@ pub async fn create(
 pub async fn get_by_id(db: &Surreal<Client>, id: &str) -> Result<Option<Submission>> {
     Ok(db.select(("submission", id)).await?)
 }
+
+pub async fn list_by_user(db: &Surreal<Client>, creator: Thing) -> Result<Vec<Submission>> {
+    Ok(db
+        .query("SELECT * FROM submission WHERE creator = $creator")
+        .bind(("creator", creator))
+        .await?
+        .take(0)?)
+}
+
+pub async fn list_by_contest(db: &Surreal<Client>, contest_id: Thing) -> Result<Vec<Submission>> {
+    Ok(db
+        .query("SELECT * FROM submission WHERE contest_id = $contest_id")
+        .bind(("contest_id", contest_id))
+        .await?
+        .take(0)?)
+}
+
+pub async fn list_within_contest(
+    db: &Surreal<Client>,
+    contest_id: Thing,
+    user_id: Thing,
+) -> Result<Vec<Submission>> {
+    let submissions = list_by_contest(db, contest_id).await?;
+
+    Ok(submissions
+        .into_iter()
+        .filter(|s| s.creator == user_id)
+        .collect())
+}
+pub async fn list_by_problem(db: &Surreal<Client>, problem_id: &str) -> Result<Vec<Submission>> {
+    Ok(db
+        .query("SELECT * FROM submission WHERE problem_id = $problem_id")
```

Comment:
Why not using `problem` instead of `problem_id`? Problem table should be `Thing`.